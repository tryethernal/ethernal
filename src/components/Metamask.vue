<template>
    <div>
        <div v-show="!ethereum">
            You need Metamask in order to interact with this contract
        </div>
        <v-row v-show="ethereum">
            <v-col>
                <v-alert outlined dense type="error" v-if="chainId && !isChainValid">
                    Invalid chain id <b>{{ chainId }}</b> ({{ parseInt(chainId, 16) }}), expecting <b>{{ formattedExpectedChainId }}</b> ({{ currentWorkspace.networkId }}). Click <a @click.stop="switchMetamaskChain()">here</a> to switch network in Metamask.
                </v-alert>
                <div v-if="connectedAccount">
                    <b>Connected Metamask account:</b> {{ connectedAccount }}
                </div>
                <v-btn :loading="loading" id="connectMetamask" v-else :color="theme == 'dark' ? '' : 'primary'" @click="connectMetamask()">Connect With Metamask</v-btn>
            </v-col>
        </v-row>
    </div>
</template>

<script>
import detectEthereumProvider from '@metamask/detect-provider';
import { mapGetters } from 'vuex';

export default {
    name: 'Metamask',
    data: () => ({
        connectedAccount: null,
        ethereum: null,
        chainId: null,
        loading: false
    }),
    mounted: function() {
        detectEthereumProvider().then((provider) => {
            if (!provider || provider !== window.ethereum) return;

            this.ethereum = provider;

            this.ethereum.on('accountsChanged', (accounts) => this.connectedAccount = accounts[0]);
            this.ethereum.on('connect', (data) => this.chainId = data.chainId);
            this.ethereum.on('chainChanged', (chainId) => this.chainId = chainId);
            this.ethereum.on('disconnect', () => {
                this.connectedAccount = null;
                this.chainId = null
            });
        });
    },
    methods: {
        connectMetamask: function() {
            this.loading = true;
            const promises = [];

            promises.push(this.ethereum.request({ method: 'eth_requestAccounts' }).then((accounts) => this.connectedAccount = accounts[0]));
            promises.push(this.ethereum.request({ method: 'eth_chainId'}).then((chainId) => this.chainId = chainId));
            Promise.all(promises).finally(() => this.loading = false);
        },
        switchMetamaskChain: function() {
            this.ethereum.request({ method: 'wallet_switchEthereumChain', params: [{ chainId: this.formattedExpectedChainId }]})
                .catch((switchError) => {
                    if (switchError.code === 4902) {
                        this.ethereum.request({
                            method: 'wallet_addEthereumChain',
                            params: [
                                {
                                    chainId: this.formattedExpectedChainId,
                                    chainName: this.publicExplorer.name,
                                    rpcUrls: [this.currentWorkspace.rpcServer],
                                    blockExplorerUrls: [this.publicExplorer.domain]
                                }
                            ]
                        }).catch(console.log);
                    }
                })
        },
        emitConnectionStatus: function() {
            const isReady = !!this.connectedAccount && this.isChainValid;
            this.$emit('rpcConnectionStatusChanged', { isReady: isReady, account: this.connectedAccount });
        }
    },
    watch: {
        'connectedAccount': function() { this.emitConnectionStatus() },
        'chainId': function() { this.emitConnectionStatus() }
    },
    computed: {
        ...mapGetters([
            'publicExplorer',
            'currentWorkspace',
            'theme'
        ]),
        isChainValid: function() {
            return this.formattedExpectedChainId === this.chainId;
        },
        formattedExpectedChainId: function() {
            return `0x${parseInt(this.currentWorkspace.networkId).toString(16)}`;
        }
    }
}
</script>
