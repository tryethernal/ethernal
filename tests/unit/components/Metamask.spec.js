import MockHelper from '../MockHelper';
import flushPromises from 'flush-promises';

import ethereum from '../mocks/ethereum';
import detectEthereumProvider from '@metamask/detect-provider';
jest.mock('@metamask/detect-provider');

import Metamask from '@/components/Metamask.vue';

describe('Metamask.vue', () => {
    let helper;

    beforeEach(async () => {
        helper = new MockHelper({
            currentWorkspace: {
                networkId: 1
            }
        });
        detectEthereumProvider.mockResolvedValue(null);
    });

    it('Should show an explaination message if no metamask', () => {
        window.ethereum = null;
        const wrapper = helper.mountFn(Metamask);
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should show a "Connect with Metamask" button if no account connected', async () => {
        window.ethereum = ethereum;
        detectEthereumProvider.mockResolvedValueOnce(window.ethereum);
        
        const wrapper = helper.mountFn(Metamask);
        await flushPromises();
        
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should show the address when an account is connected', async () => {
        jest.spyOn(ethereum, 'request')
            .mockResolvedValueOnce(['0x1234'])
            .mockResolvedValueOnce('0x1');
        window.ethereum = ethereum;
        detectEthereumProvider.mockResolvedValueOnce(window.ethereum);

        const wrapper = helper.mountFn(Metamask);
        await flushPromises();

        await wrapper.find('#connectMetamask').trigger('click');
        await flushPromises();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should show an error is Metamask is not on the expected network', async () => {
        jest.spyOn(ethereum, 'request')
            .mockResolvedValueOnce(['0x1234'])
            .mockResolvedValueOnce('0x2');
        window.ethereum = ethereum;
        detectEthereumProvider.mockResolvedValueOnce(window.ethereum);

        const wrapper = helper.mountFn(Metamask);
        await flushPromises();

        await wrapper.find('#connectMetamask').trigger('click');
        await flushPromises();

        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should emit the new connection status when connecting Metamask', async () => {
        jest.spyOn(ethereum, 'request')
            .mockResolvedValueOnce(['0x1234'])
            .mockResolvedValueOnce('0x1');
        window.ethereum = ethereum;
        detectEthereumProvider.mockResolvedValueOnce(window.ethereum);

        const wrapper = helper.mountFn(Metamask);
        await flushPromises();
        
        await wrapper.find('#connectMetamask').trigger('click');
        await flushPromises();

        expect(wrapper.emitted().rpcConnectionStatusChanged.length).toBe(2);
        expect(wrapper.vm.connectedAccount).toEqual('0x1234');
        expect(wrapper.vm.chainId).toEqual('0x1');
    });
});
