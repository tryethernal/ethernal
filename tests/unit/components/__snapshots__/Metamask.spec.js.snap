// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Metamask.vue Should show a "Connect with Metamask" button if no account connected 1`] = `
<div>
  <div style="display: none;">
    You need Metamask in order to interact with this contract
  </div>
  <div class="row" style="">
    <div class="col">
      <!----> <button type="button" class="v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default primary" id="connectMetamask"><span class="v-btn__content">Connect With Metamask</span></button>
    </div>
  </div>
</div>
`;

exports[`Metamask.vue Should show an error is Metamask is not on the expected network 1`] = `
<div>
  <div style="display: none;">
    You need Metamask in order to interact with this contract
  </div>
  <div class="row" style="">
    <div class="col">
      <div role="alert" class="v-alert v-sheet v-sheet--outlined theme--light v-alert--dense v-alert--outlined error--text">
        <div class="v-alert__wrapper"><i aria-hidden="true" class="v-icon notranslate v-alert__icon mdi mdi-alert theme--light error--text"></i>
          <div class="v-alert__content">
            Invalid chain id <b>0x2</b> (2), expecting <b>0x1</b> (1). Click <a>here</a> to switch network in Metamask.
          </div>
        </div>
      </div>
      <div><b>Connected Metamask account:</b> 0x1234
      </div>
    </div>
  </div>
</div>
`;

exports[`Metamask.vue Should show an explaination message if no metamask 1`] = `
<div>
  <div>
    You need Metamask in order to interact with this contract
  </div>
  <div class="row" style="display: none;">
    <div class="col">
      <!----> <button type="button" class="v-btn v-btn--is-elevated v-btn--has-bg theme--light v-size--default primary" id="connectMetamask"><span class="v-btn__content">Connect With Metamask</span></button>
    </div>
  </div>
</div>
`;

exports[`Metamask.vue Should show the address when an account is connected 1`] = `
<div>
  <div style="display: none;">
    You need Metamask in order to interact with this contract
  </div>
  <div class="row" style="">
    <div class="col">
      <!---->
      <div><b>Connected Metamask account:</b> 0x1234
      </div>
    </div>
  </div>
</div>
`;
